Python/import.c

```diff
@@ -1369,11 +1411,21 @@ import_find_extension(PyThreadState *tstate,
         }
         struct _Py_ext_module_loader_result res;
         if (_PyImport_RunModInitFunc(def->m_base.m_init, info, &res) < 0) {
+            _Py_ext_module_loader_result_apply_error(&res, name_buf);
+            _Py_ext_module_loader_result_clear(&res);
             return NULL;
         }
-        assert(!PyErr_Occurred());
+        assert(!PyErr_Occurred() && res.err == NULL);
+        assert(res.kind == _Py_ext_module_kind_SINGLEPHASE);
         mod = res.module;
+        /* Tchnically, the init function could return a different module def.
+         * Then we would probably need to update the global cache.
+         * However, we don't expect anyone to change the def. */
+        assert(res.def == def);
+        _Py_ext_module_loader_result_clear(&res);
+
         // XXX __file__ doesn't get set!
```

Comment:
This is now solved, __file__ gets set elsewhere.
```suggestion

```

Reply:
While the import machinery (in `importlib._bootstrap`) *does* set `__file__`, the only place it gets set for a single-phase init module via `_imp.load_dynamic` (ergo `ExtensionFileLoader.exec_module()`) is the first time it is loaded, in `import_run_extension()`.  In the above case, the module was found in the global cache and gets reloaded from the cached init function.

I was going to clarify the comment, but it would be just as easy to actually set `__file__` and be done with it. 😄

Reply:
done